File: android/src/main/java/com/hypertrack/sdk/capacitor/HyperTrackSdkPlugin.java
package com.hypertrack.sdk.capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.hypertrack.sdk.AvailabilityError;
import com.hypertrack.sdk.AvailabilityStateObserver;
import com.hypertrack.sdk.TrackingError;
import com.hypertrack.sdk.TrackingStateObserver;

import org.json.JSONArray;
import org.json.JSONObject;

@CapacitorPlugin(name = "HyperTrackSdk")
public class HyperTrackSdkPlugin extends Plugin implements TrackingStateObserver.OnTrackingStateChangeListener, AvailabilityStateObserver.OnAvailabilityStateChangeListener {

    private final HyperTrackSdk implementation = new HyperTrackSdk();

    @PluginMethod
    public void initialize(PluginCall call) {
        implementation.print("initialize called");
        try {
            String publishableKey = call.getString("publishableKey");
            implementation.print(publishableKey);
            implementation.initialize(publishableKey);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void start(PluginCall call) {
        implementation.print("start called");
        try {
            implementation.start();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void stop(PluginCall call) {
        implementation.print("stop called");
        try {
            implementation.stop();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void enableDebugLogging(PluginCall call) {
        implementation.print("enableDebugLogging called");
        try {
            implementation.enableDebugLogging();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void requestPermissionsIfNecessary(PluginCall call) {
        implementation.print("requestPermissionsIfNecessary called");
        try {
            implementation.requestPermissionsIfNecessary();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void allowMockLocations(PluginCall call) {
        implementation.print("allowMockLocations called");
        try {
            implementation.hyperTrackAllowMockLocations();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void getDeviceId(PluginCall call) {
        implementation.print("getDeviceId called");
        try {
            String deviceId = implementation.getDeviceId();
            JSObject result = new JSObject();
            result.put("deviceId", implementation.print(deviceId));
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void setDeviceName(PluginCall call) {
        implementation.print("setDeviceName called");
        String deviceName = call.getString("name");
        try {
            implementation.setDeviceName(deviceName);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod(returnType = PluginMethod.RETURN_CALLBACK)
    public void getLatestLocation(PluginCall call) {
        implementation.print("getLatestLocation called");
        try {
            JSONObject latestLocation = implementation.getLatestLocation().getJSONObject("location");
            JSObject result = new JSObject();
            result.put("location", latestLocation);
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void addGeotag(PluginCall call) {
        implementation.print("addGeotag called");
        try {
            String geoTagData = call.getData().getString("metadata");
            String expectedLocation = call.getData().getString("coordinates");
            implementation.addGeotag(geoTagData, expectedLocation);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void setDeviceMetadata(PluginCall call) {
        implementation.print("setDeviceMetadata called");
        try {
            JSObject deviceMetaJsonObject = call.getData();
            implementation.setDeviceMetadata(deviceMetaJsonObject);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void syncDeviceSettings(PluginCall call) {
        implementation.print("syncDeviceSettings called");
        try {
            implementation.syncDeviceSettings();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void getAvailability(PluginCall call) {
        implementation.print("getAvailability called");
        try {
            Object availability = implementation.getAvailability();
            JSObject result = new JSObject();
            result.put("status", availability);
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void setAvailability(PluginCall call) {
        implementation.print("setAvailability called");
        try {
            Boolean isAvailable = call.getBoolean("isAvailable");
            implementation.setAvailability(isAvailable);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void isTracking(PluginCall call) {
        implementation.print("isTracking called");
        try {
            boolean isTracking = implementation.isTracking();
            JSObject result = new JSObject();
            result.put("status", isTracking);
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void addTrackingListener(PluginCall call) {
        implementation.print("addTrackingListener called");
        try {
            implementation.addTrackingListener(this);
            implementation.print("Tracker listener added successfully.");
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void removeTrackingListener(PluginCall call) {
        implementation.print("removeTrackingListener called");
        try {
            implementation.removeTrackingListener(this);
            implementation.print("Tracker listener removed successfully.");
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void setTrackingNotificationProperties(PluginCall call) {
        implementation.print("setTrackingNotificationProperties called");
        try {
            String title = call.getData().getString("title");
            String body = call.getData().getString("message");
            implementation.setTrackingNotificationProperties(title, body);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void addAvailabilityListener(PluginCall call) {
        implementation.print("addAvailabilityListener called");
        try {
            implementation.addAvailabilityListener(this);
            implementation.print("Availability listener added successfully.");
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void removeAvailabilityListener(PluginCall call) {
        implementation.print("removeAvailabilityListener called");
        try {
            implementation.removeAvailabilityListener(this);
            implementation.print("Availability listener removed successfully.");
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void getBlockers(PluginCall call) {
        implementation.print("getBlockers called");
        try {
            JSONArray blockers = implementation.getBlockers();
            JSObject result = new JSObject();
            result.put("blockers", blockers);
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @PluginMethod
    public void resolveBlocker(PluginCall call) {
        implementation.print("resolveBlocker called");
        String code = call.getString("code");
        try {
            implementation.resolveBlocker(code);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.toString(), e);
        }
    }

    @Override
    public void onError(TrackingError trackingError) {
        JSObject result = new JSObject();
        result.put("error", trackingError.toString());
        notifyListeners("trackingStateChange", result);
    }

    @Override
    public void onTrackingStart() {
        JSObject result = new JSObject();
        result.put("status", "start");
        notifyListeners("trackingStateChange", result);
    }

    @Override
    public void onTrackingStop() {
        JSObject result = new JSObject();
        result.put("status", "stop");
        notifyListeners("trackingStateChange", result);
    }

    @Override
    public void onError(AvailabilityError availabilityError) {
        JSObject result = new JSObject();
        result.put("error", availabilityError.toString());
        notifyListeners("availabilityStateChange", result);
    }

    @Override
    public void onAvailable() {
        JSObject result = new JSObject();
        result.put("availability", true);
        notifyListeners("availabilityStateChange", result);
    }

    @Override
    public void onUnavailable() {
        JSObject result = new JSObject();
        result.put("availability", false);
        notifyListeners("availabilityStateChange", result);
    }
}

File: ios/Plugin/HyperTrackSdkPlugin.swift
import Foundation
import Capacitor
import HyperTrack
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(HyperTrackSdkPlugin)
public class HyperTrackSdkPlugin: CAPPlugin {
    private let implementation = HyperTrackSdk()
    
    private var startedTrackingNotificationObserver: Any? = nil
    private var stoppedTrackingNotificationObserver: Any? = nil
    private var didEncounterUnrestorableErrorNotificationObserver: Any? = nil
    private var didEncounterRestorableErrorNotificationObserver: Any? = nil
    private var becameAvailableNotificationObserver: Any? = nil
    private var becameUnavailableNotificationObserver: Any? = nil
    
    @objc func addTrackingListener() {
        startedTrackingNotificationObserver = NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.startTracking),
            name: HyperTrack.startedTrackingNotification,
            object: nil)
        
        stoppedTrackingNotificationObserver =  NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.stopTracking),
            name: HyperTrack.stoppedTrackingNotification,
            object: nil
        )
        
        didEncounterUnrestorableErrorNotificationObserver = NotificationCenter.default.addObserver(
            self,
            selector: #selector(trackingError(notification:)),
            name: HyperTrack.didEncounterUnrestorableErrorNotification,
            object: nil
        )
        didEncounterRestorableErrorNotificationObserver = NotificationCenter.default.addObserver(
            self,
            selector: #selector(trackingError(notification:)),
            name: HyperTrack.didEncounterRestorableErrorNotification,
            object: nil
        )
    }
    
    @objc func removeTrackingListener() {
        NotificationCenter.default.removeObserver(startedTrackingNotificationObserver!)
        NotificationCenter.default.removeObserver(stoppedTrackingNotificationObserver!)
        NotificationCenter.default.removeObserver(didEncounterUnrestorableErrorNotificationObserver!)
        NotificationCenter.default.removeObserver(didEncounterRestorableErrorNotificationObserver!)
    }
    
    @objc func addAvailabilityListener() {
        becameAvailableNotificationObserver = NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.available),
            name: HyperTrack.becameAvailableNotification,
            object: nil)
        
        becameUnavailableNotificationObserver =  NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.unavailable),
            name: HyperTrack.becameUnavailableNotification,
            object: nil
        )
    }
    
    @objc func removeAvailabilityListener() {
        NotificationCenter.default.removeObserver(becameAvailableNotificationObserver!)
        NotificationCenter.default.removeObserver(becameUnavailableNotificationObserver!)
    }
    
    @objc private func trackingError(notification: Notification) {
        if let trackingError = notification.hyperTrackTrackingError() {
            if let restorableError = trackingError.restorableError {
                notifyListeners("trackingStateChange", data: ["error" : restorableError.localizedDescription])
            }
            if let unrestorableError = trackingError.unrestorableError {
                notifyListeners("trackingStateChange", data: ["error" : unrestorableError.localizedDescription])
            }
        }
    }
    
    @objc private func startTracking() {
        notifyListeners("trackingStateChange", data: ["status" : "start"])
    }
    
    @objc private func stopTracking() {
        notifyListeners("trackingStateChange", data: ["status" : "stop"])
    }
    
    @objc private func available() {
        notifyListeners("availabilityStateChange", data: ["availability" : true])
    }
    
    @objc private func unavailable() {
        notifyListeners("availabilityStateChange", data: ["availability" : false])
    }
    
    @objc func initialize(_ call: CAPPluginCall) {
        let publishableKey = call.getString("publishableKey") ?? ""
        do {
            try implementation.initialize(publishableKey: publishableKey)
            call.resolve()
        } catch let error as NSError {
            call.reject(error.localizedDescription, nil, error)
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func addGeotag(_ call: CAPPluginCall) {
        let tag = call.getString("metadata") ?? ""
        do {
            try implementation.addGeotag(tag: tag)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func setDeviceMetadata(_ call: CAPPluginCall) {
        let metaData = call.getString("metaData") ?? ""
        do {
            try implementation.setDeviceMetadata(metaData: metaData)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func isTracking(_ call: CAPPluginCall) {
        do {
            let status = try implementation.isTracking()
            call.resolve(["status":status])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func syncDeviceSettings(_ call: CAPPluginCall) {
        do {
            try implementation.syncDeviceSettings()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func allowMockLocations(_ call: CAPPluginCall) {
        do {
            try implementation.allowMockLocations()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func start(_ call: CAPPluginCall) {
        do {
            try implementation.start()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func stop(_ call: CAPPluginCall) {
        do {
            try implementation.stop()
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func setDeviceName(_ call: CAPPluginCall) {
        let deviceName = call.getString("name") ?? ""
        do {
            try implementation.setDeviceName(deviceName: deviceName)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func getAvailability(_ call: CAPPluginCall) {
        do {
            let status = try implementation.availability()
            call.resolve(["status":status])
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func enableDebugLogging(_ call: CAPPluginCall) {
        implementation.enableDebugLogging()
        call.resolve()
    }
}
